File: day-2/main.py
#!/usr/bin/env python3

from os import PathLike

type Report = list[int]


def parse_input(filepath: str | PathLike[str]) -> list[Report]:
    reports = []

    with open(filepath, "r") as f:
        for line in f:
            report = list(map(int, line.split()))
            reports.append(report)

    return reports


def is_monotonic(report: Report) -> bool:
    return report == sorted(report) or report == sorted(report, reverse=True)


def has_gradual_evolution(report: Report) -> bool:
    for i, elm in enumerate(report):
        if i == 0:
            continue

        level_diff = elm - report[i - 1]
        if abs(level_diff) < 1 or abs(level_diff) > 3:
            return False

    return True


def is_safe_report(report: Report) -> bool:
    return is_monotonic(report) and has_gradual_evolution(report)


def count_safe_reports(reports: list[Report], tolerant: bool = False) -> int:
    safe_reports = 0
    for report in reports:
        if is_safe_report(report):
            safe_reports += 1
            continue

        if not tolerant:
            continue

        for i, _ in enumerate(report):
            cutted_report = report.copy()
            cutted_report.pop(i)

            if is_safe_report(cutted_report):
                safe_reports += 1
                break

    return safe_reports


def main():
    reports = parse_input("input.txt")

    safe_reports = count_safe_reports(reports)

    print(f"Safe reports: {safe_reports}")

    tolerated_reports = count_safe_reports(reports, True)

    print(f"Tolerated reports: {tolerated_reports}")


if __name__ == "__main__":
    main()

File: day-2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Report []int

func stringToIntegerSlice(line []string) (Report, error) {
	var numbers []int
	for _, elm := range line {
		number, err := strconv.Atoi(elm)
		if err != nil {
			return numbers, err
		}

		numbers = append(numbers, number)
	}

	return numbers, nil
}

func parseInput(filepath string) ([]Report, error) {
	file, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}

	defer file.Close()

	var reports []Report

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.Fields(scanner.Text())
		report, err := stringToIntegerSlice(line)
		if err != nil {
			return nil, err
		}
		reports = append(reports, report)
	}
	return reports, nil
}

func isMonotonic(report Report) bool {
	reversed := slices.Clone(report)
	slices.Reverse(reversed)
	return slices.IsSorted(report) || slices.IsSorted(reversed)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func hasGradualEvolution(report Report) bool {
	for i := 1; i < len(report); i++ {
		levelDiff := report[i] - report[i-1]
		if abs(levelDiff) < 1 || abs(levelDiff) > 3 {
			return false
		}
	}

	return true
}

func isSafeReport(report Report) bool {
	return isMonotonic(report) && hasGradualEvolution(report)
}

func countSafeReport(reports []Report, tolerant bool) int {
	safeReports := 0
	for _, report := range reports {
		if isSafeReport(report) {
			safeReports++
			continue
		}

		if !tolerant {
			continue
		}

		for i := range report {
			cuttedReport := slices.Clone(report)
			cuttedReport = append(cuttedReport[:i], cuttedReport[i+1:]...)

			if isSafeReport(cuttedReport) {
				safeReports++
				break
			}
		}
	}

	return safeReports
}

func main() {
	reports, err := parseInput("input.txt")
	if err != nil {
		panic(err)
	}

	safeReports := countSafeReport(reports, false)

	fmt.Println("Safe reports:", safeReports)

	toleratedReports := countSafeReport(reports, true)

	fmt.Println("Tolerated reports:", toleratedReports)
}
